src/Microsoft.AspNetCore.Cors/Infrastructure/CorsService.cs

```diff
@@ -125,40 +125,32 @@ public virtual void EvaluatePreflightRequest(HttpContext context, CorsPolicy pol
             var requestHeaders =
                 context.Request.Headers.GetCommaSeparatedValues(CorsConstants.AccessControlRequestHeaders);
 
-            if (!policy.AllowAnyMethod)
+            if (!policy.AllowAnyMethod && !policy.Methods.Contains(accessControlRequestMethod.ToString(), StringComparer.OrdinalIgnoreCase))
             {
-                var found = false;
-                for (var i = 0; i < policy.Methods.Count; i++)
-                {
-                    var method = policy.Methods[i];
-                    if (string.Equals(method, accessControlRequestMethod, StringComparison.OrdinalIgnoreCase))
-                    {
-                        found = true;
-                        break;
-                    }
-                }
-
-                if (!found)
-                {
-                    _logger?.PolicyFailure();
-                    _logger?.AccessControlMethodNotAllowed(accessControlRequestMethod);
-                    return;
-                }
+                _logger?.PolicyFailure();
+                _logger?.AccessControlMethodNotAllowed(accessControlRequestMethod);
+                return;
             }
 
-            if (!policy.AllowAnyHeader &&
-                requestHeaders != null)
+            if (requestHeaders != null)
             {
-                foreach (var requestHeader in requestHeaders)
+                if (!policy.AllowAnyHeader)
                 {
-                    if (!CorsConstants.SimpleRequestHeaders.Contains(requestHeader, StringComparer.OrdinalIgnoreCase) &&
-                                                  !policy.Headers.Contains(requestHeader, StringComparer.OrdinalIgnoreCase))
+                    foreach (var requestHeader in requestHeaders)
                     {
-                        _logger?.PolicyFailure();
-                        _logger?.RequestHeaderNotAllowed(requestHeader);
-                        return;
+                        if (!CorsConstants.SimpleRequestHeaders.Contains(requestHeader, StringComparer.OrdinalIgnoreCase) &&
+                                                      !policy.Headers.Contains(requestHeader, StringComparer.OrdinalIgnoreCase))
+                        {
+                            _logger?.PolicyFailure();
+                            _logger?.RequestHeaderNotAllowed(requestHeader);
+                            return;
+                        }
                     }
                 }
+
+                result.FilterSimpleMethods
+                    = !(CorsConstants.SimpleMethods.Contains(accessControlRequestMethod.ToString(), StringComparer.OrdinalIgnoreCase) &&
+                        requestHeaders.Contains("content-type", StringComparer.OrdinalIgnoreCase));
```

Comment:
Shouldn't we check here for the content-types we know shouldn't trigger a preflight?
```
application/x-www-form-urlencoded
multipart/form-data
text/plain
```

Reply:
Preflight requests do not send this information.
As per spec the client is the one enforcing this check, if it doesn't use one of those 3 mime-types it'll do a request using the Access-Control-Allow-Headers: content-type header, not specifying the content-type because it will do that in the actual request.

Reply:
@FlyingWraptor Thanks for the clarification. You are totally right.